### ISSUE TITLE
Any advice if gradient explodes?

### ISSUE THREAD

**Issue by ShiCrazy** [vkola-lab/tmi2022#7]
I print 'loss' and 'out', and I find many "nan" during the training process. It seems that gradient explodes. I tried reducing learning rate but still failed to avoid the problem. Do you have any advice? Thank you in advance!

**Comment by ZetaLx** [comment]
> I print 'loss' and 'out', and I find many "nan" during the training process. It seems that gradient explodes. I tried reducing learning rate but still failed to avoid the problem. Do you have any advice? Thank you in advance!

May I ask how you made the all_patches.CSV file? Thank you in advance!

**Comment by ShiCrazy** [comment]
> > I print 'loss' and 'out', and I find many "nan" during the training process. It seems that gradient explodes. I tried reducing learning rate but still failed to avoid the problem. Do you have any advice? Thank you in advance!
> 
> May I ask how you made the all_patches.CSV file? Thank you in advance!

I did not make a "all_patches.csv" file because the framework do not need it. I just made a series of folders like "SLIDES", "PATCHES", "GRAPHS" to store my WSI slides, the patches generated and something else. You have to also prepare a "train_set.txt", "val_set.txt"  and "test_set.txt".

What you need to do is just follow the "README" file provided by the author and then you can train a model.

**Comment by ZetaLx** [comment]
> 

So you just used his model structure and not his github code? Because I saw that the dataset in the pre-training part is read by reading the csv file, or did you modify the trainset part?

**Comment by ShiCrazy** [comment]
> > 
> 
> So you just used his model structure and not his github code? Because I saw that the dataset in the pre-training part is read by reading the csv file, or did you modify the trainset part?

Making a "all_patches.csv" is necessary only when you want to train a **feature extractor** by yourself.

**Comment by ZetaLx** [comment]
> > > 
> > 
> > 
> > So you just used his model structure and not his github code? Because I saw that the dataset in the pre-training part is read by reading the csv file, or did you modify the trainset part?
> 
> Making a "all_patches.csv" is necessary only when you want to train a **feature extractor** by yourself.

Yes, since I want to use this model on another dataset, I want to do the self-supervised part of the pre-training myself. But I don't know how to style and make csv. Can you give me some advice if you know? Thank you so much!

**Comment by ShiCrazy** [comment]
> Yes, since I want to use this model on another dataset, I want to do the self-supervised part of the pre-training myself. But I don't know how to style and make csv. Can you give me some advice if you know? Thank you so much!

you can find a class called "Dataset()" in dataset_wrapper.py. Accroding to the codes, it seems you need to put all paths of you patches generated from your slides in the first column of the csv file

**Comment by ZetaLx** [comment]
> > Yes, since I want to use this model on another dataset, I want to do the self-supervised part of the pre-training myself. But I don't know how to style and make csv. Can you give me some advice if you know? Thank you so much!
> 
> you can find a class called "Dataset()" in dataset_wrapper.py. Accroding to the codes, it seems you need to put all paths of you patches generated from your slides in the first column of the csv file

Thank you very much!

**Comment by ChenYuhang243** [comment]
> > > Yes, since I want to use this model on another dataset, I want to do the self-supervised part of the pre-training myself. But I don't know how to style and make csv. Can you give me some advice if you know? Thank you so much!
> > 
> > 
> > you can find a class called "Dataset()" in dataset_wrapper.py. Accroding to the codes, it seems you need to put all paths of you patches generated from your slides in the first column of the csv file
> 
> Thank you very much!
following codes copied directly from https://github.com/binli123/dsmil-wsi/simclr/run.py
`
def generate_csv(args):
    if args.level=='high' and args.multiscale==1:
        path_temp = os.path.join('..', 'WSI', args.dataset, 'pyramid', '*', '*', '*', '*.jpeg')
        patch_path = glob.glob(path_temp) # /class_name/bag_name/5x_name/*.jpeg
    if args.level=='low' and args.multiscale==1:
        path_temp = os.path.join('..', 'WSI', args.dataset, 'pyramid', '*', '*', '*.jpeg')
        patch_path = glob.glob(path_temp) # /class_name/bag_name/*.jpeg
    if args.multiscale==0:
        path_temp = os.path.join('..', 'WSI', args.dataset, 'single', '*', '*', '*.jpeg')
        patch_path = glob.glob(path_temp) # /class_name/bag_name/*.jpeg
    df = pd.DataFrame(patch_path)
    df.to_csv('all_patches.csv', index=False)
`

**Comment by ZetaLx** [comment]
> > > > Yes, since I want to use this model on another dataset, I want to do the self-supervised part of the pre-training myself. But I don't know how to style and make csv. Can you give me some advice if you know? Thank you so much!
> > > 
> > > 
> > > you can find a class called "Dataset()" in dataset_wrapper.py. Accroding to the codes, it seems you need to put all paths of you patches generated from your slides in the first column of the csv file
> > 
> > 
> > Thank you very much!
> > following codes copied directly from https://github.com/binli123/dsmil-wsi/simclr/run.py
> > `def generate_csv(args):     if args.level=='high' and args.multiscale==1:         path_temp = os.path.join('..', 'WSI', args.dataset, 'pyramid', '*', '*', '*', '*.jpeg')         patch_path = glob.glob(path_temp) # /class_name/bag_name/5x_name/*.jpeg     if args.level=='low' and args.multiscale==1:         path_temp = os.path.join('..', 'WSI', args.dataset, 'pyramid', '*', '*', '*.jpeg')         patch_path = glob.glob(path_temp) # /class_name/bag_name/*.jpeg     if args.multiscale==0:         path_temp = os.path.join('..', 'WSI', args.dataset, 'single', '*', '*', '*.jpeg')         patch_path = glob.glob(path_temp) # /class_name/bag_name/*.jpeg     df = pd.DataFrame(patch_path)     df.to_csv('all_patches.csv', index=False)`

Thank you so much for the code!
